RooFit::MultiProcess & TestStatistics part 4: RooMinimizerFcn refactoring

egpbos opened the pull request:
This PR refactors RooMinimizerFcn into two separate classes: RooAbsMinimizerFcn and a new RooMinimizerFcn derived from that. RooAbsMinimizerFcn contains all common synchronization "logistics" necessary between RooFit and Minuit function parameters. The implementations of this abstract base class are then expected to supply the actual function evaluation parts. This allows us to later implement classes that calculate the gradient outside of Minuit. We will provide both a serial implementation of this and later on a parallel one using the MultiProcess framework.

## Checklist:

- [ ] tested changes locally
- [ ] updated the docs (if necessary)

guitargeek commented on roofit/roofitcore/src/RooAbsMinimizerFcn.cxx:
```diff
@@ -0,0 +1,526 @@
+ * listed in LICENSE (http://roofit.sourceforge.net/license.txt)             *
+ *****************************************************************************/
+
+#ifndef __ROOFIT_NOROOMINIMIZER
```
Is this `ifndef` really needed? When would we build with `__ROOFIT_NOROOMINIMIZER`?

egpbos replied:
It was already there, but seems unused. Wouter confirms that it is obsolete, so could be removed in all places that it is still used. Separate PR probably?

guitargeek replied:
Ah sorry, I didn't realize it was there before. I agree, everything that is cleaning up old stuff is a task for a separate PR.

hageboeck replied:
No, not needed. Agree about a separate PR, where things like [this part here](https://github.com/root-project/root/blob/7c4f9321a3fcbda598da6b815409a9a5ed72a53e/roofit/roofitcore/src/RooAbsPdf.cxx#L1599-L1783) could be condensed down into the one branch that's actually used.

egpbos replied:
I will just remove them when I encounter them.

egpbos replied:
Oh weird, I didn't see your replies when I typed that I would just remove them, should have refreshed the page. I didn't actually notice the part that you linked @hageboeck. The ifdef's I encountered so far were completely inconsequential (those that switch on/off the entire file, which I think really is useless), so perhaps those can be safely removed while the one you mention would need a little extra motivation and double checking.

hageboeck replied:
Yes, please do remove them here, and Jonas can take care of the ones that are left in roofitcore.
#8604

guitargeek commented on roofit/roofitcore/src/RooMinimizer.cxx:
```diff
@@ -959,5 +959,9 @@ RooFitResult* RooMinimizer::lastMinuitFit(const RooArgList& varList)
   return res;
 
 }
+Int_t RooMinimizer::getPrintLevel() const
```
I would not outline these one-liners. They should better be inlined in the header file. There are also many such one-liners in RooAbsMininmizerFcn, but since they are copy pasted from the original file, I'm fine with keeping them there. But for new code, it would be nice to inline.

guitargeek commented on roofit/roofitcore/src/RooMinimizerFcn.cxx:
```diff
@@ -541,4 +149,14 @@ double RooMinimizerFcn::DoEval(const double *x) const {
   return fvalue;
 }
 
+std::string RooMinimizerFcn::getFunctionName() const
```
These two functions `getFunctionName` and `getFunctionTitle` can also be inlined in the header to allow for better compiler optimization and improve code readability.

hageboeck replied:
Maybe, maybe not. I left some comments to bring down the includes in the headers. If you follow that route, it's actually better to forward declare `RooAbsReal`, and have those functions in the `cxx`.
If `RooAbsReal` needs to be included in the header, I would suggest inlining.

guitargeek replied:
Good point, if inlining means to add mode includes that would otherwise not be necessary it should not be done.

guitargeek commented on roofit/roofitcore/inc/RooMinimizerFcn.h:
```diff
@@ -26,83 +27,34 @@
-  RooAbsReal *_funct;
-  const RooMinimizer *_context;
+   ROOT::Math::IBaseFunctionMultiDim *Clone() const override;
+   unsigned int NDim() const override { return get_nDim(); }
```
`get_nDim` is not RooFit variable naming convention with the underscore in the name. `getNDim` would be more consistent.

hageboeck commented on roofit/roofitcore/inc/RooMinimizerFcn.h:
```diff
@@ -26,83 +27,34 @@
+class RooMinimizer;
 
-  virtual double DoEval(const double * x) const;  
+class RooMinimizerFcn : public RooAbsMinimizerFcn, public ROOT::Math::IBaseFunctionMultiDim {
```
Multiple inheritance from classes that are not abstract isn't nice in C++. Now all the stuff such as having two separate data sections, pointer changes on casting etc are going to happen.
If possible, make `AbsMinimizerFcn` really abstract, i.e. **no** members.

hageboeck commented on roofit/roofitcore/inc/RooMinimizerFcn.h:
```diff
@@ -26,83 +27,34 @@
-  void zeroEvalCount() { _evalCounter = 0 ; }
-  /// Return a possible offset that's applied to the function to separate invalid function values from valid ones.
-  double getOffset() const { return _funcOffset; }
-
- private:
-  void SetPdfParamErr(Int_t index, Double_t value);
-  void ClearPdfParamAsymErr(Int_t index);
-  void SetPdfParamErr(Int_t index, Double_t loVal, Double_t hiVal);
-
-  Bool_t SetPdfParamVal(int index, double value) const;
-  void printEvalErrors() const;
+// forward declaration
+class RooMinimizer;
```
Those forward declarations are on honourable attempt, but since the abstract version includes all those things, they are unnecessary.
To detect include errors in other classes, the best ordering is
```c++
#include "MyBaseClass.h"

#include "RooFitStuff.h"
#include "RooFitStuff2.h"

#include "RootStuff.h"

#include <stdlib>
```

But again: Forward declare if possible.

hageboeck commented on roofit/roofitcore/inc/RooAbsMinimizerFcn.h:
```diff
@@ -0,0 +1,131 @@
+   RooAbsMinimizerFcn(const RooAbsMinimizerFcn &other);
+   virtual ~RooAbsMinimizerFcn();
+
+   // inform Minuit through its parameter_settings vector of RooFit parameter properties
+   Bool_t synchronize_parameter_settings(std::vector<ROOT::Fit::ParameterSettings> &parameters, Bool_t optConst, Bool_t verbose);
```
Naming conventions? Add doxygen documentation, please.

hageboeck commented on roofit/roofitcore/inc/RooAbsMinimizerFcn.h:
```diff
@@ -0,0 +1,131 @@
+
+   // inform Minuit through its parameter_settings vector of RooFit parameter properties
+   Bool_t synchronize_parameter_settings(std::vector<ROOT::Fit::ParameterSettings> &parameters, Bool_t optConst, Bool_t verbose);
+   // same, but can be overridden to e.g. also include gradient strategy synchronization in subclasses:
+   virtual Bool_t Synchronize(std::vector<ROOT::Fit::ParameterSettings> &parameters, Bool_t optConst, Bool_t verbose);
+
+   // used for export to RooFitResult from Minimizer:
+   RooArgList *GetFloatParamList();
+   RooArgList *GetConstParamList();
+   RooArgList *GetInitFloatParamList();
+   RooArgList *GetInitConstParamList();
+   Int_t GetNumInvalidNLL() const;
```
Doxygen also here.

hageboeck commented on roofit/roofitcore/inc/RooAbsMinimizerFcn.h:
```diff
@@ -0,0 +1,131 @@
+   RooArgList *GetInitConstParamList();
+   Int_t GetNumInvalidNLL() const;
+
+   // need access from Minimizer:
+   void SetEvalErrorWall(Bool_t flag);
```
Doxygen.

hageboeck commented on roofit/roofitcore/inc/RooAbsMinimizerFcn.h:
```diff
@@ -0,0 +1,131 @@
+   void SetRecoverFromNaNStrength(double strength) { _recoverFromNaNStrength = strength; }
+   void SetPrintEvalErrors(Int_t numEvalErrors);
+   Double_t &GetMaxFCN();
+   Int_t evalCounter() const;
+   void zeroEvalCount();
+   /// Return a possible offset that's applied to the function to separate invalid function values from valid ones.
+   double getOffset() const { return _funcOffset; }
+   void SetVerbose(Bool_t flag = kTRUE);
+
+   // put Minuit results back into RooFit objects:
+   void BackProp(const ROOT::Fit::FitResult &results);
+
+   // used in several Minimizer functions:
+   virtual std::string getFunctionName() const = 0;
+   virtual std::string getFunctionTitle() const = 0;
+
+   // set different external covariance matrix
+   void ApplyCovarianceMatrix(TMatrixDSym &V);
+
+   Bool_t SetLogFile(const char *inLogfile);
+   std::ofstream *GetLogFile() { return _logfile; }
+
+   unsigned int get_nDim() const { return _nDim; }
+
+   virtual void setOptimizeConst(Int_t flag) = 0;
+
+   bool getOptConst();
+   std::vector<double> get_parameter_values() const;
+
+   Bool_t SetPdfParamVal(int index, double value) const;
+
+protected:
+   virtual void optimizeConstantTerms(bool constStatChange, bool constValChange) = 0;
+
+   // used in BackProp (Minuit results -> RooFit) and ApplyCovarianceMatrix
+   void SetPdfParamErr(Int_t index, Double_t value);
+   void ClearPdfParamAsymErr(Int_t index);
+   void SetPdfParamErr(Int_t index, Double_t loVal, Double_t hiVal);
+
+   void printEvalErrors() const;
```
Doxygen also here. This is especially important since this class defines the interface, so users (implementers) should be able to look up what the functions are supposed to do.

hageboeck commented on roofit/roofitcore/inc/RooAbsMinimizerFcn.h:
```diff
@@ -0,0 +1,131 @@
+
+   // the following four are mutable because DoEval is const (in child classes)
+   // Reset the *largest* negative log-likelihood value we have seen so far:
+   mutable double _maxFCN = -std::numeric_limits<double>::infinity();
+   mutable double _funcOffset{0.};
+   double _recoverFromNaNStrength{10.};
+   mutable int _numBadNLL = 0;
+   mutable int _printEvalErrors = 10;
+   mutable int _evalCounter{0};
+
+   unsigned int _nDim = 0;
+
+   Bool_t _optConst = kFALSE;
+
+   RooArgList *_floatParamList;
+   RooArgList *_constParamList;
+   RooArgList *_initFloatParamList;
+   RooArgList *_initConstParamList;
+
+   std::ofstream *_logfile = nullptr;
+   bool _doEvalErrorWall{true};
+   bool _verbose;
```
Try not to have data members in abstract classes.
If there's no way around them, better make them private. Making them protected increases the amount of code that has access to them, and which will make refactoring harder, since there's more code that has to be touched.

hageboeck commented on roofit/roofitcore/inc/RooAbsMinimizerFcn.h:
```diff
@@ -0,0 +1,131 @@
+
+   Bool_t _optConst = kFALSE;
+
+   RooArgList *_floatParamList;
+   RooArgList *_constParamList;
+   RooArgList *_initFloatParamList;
+   RooArgList *_initConstParamList;
```
When touching this anyway: `unique_ptr`?

hageboeck reviewed (commented):
I have some suggestions regarding keeping headers small and with few dependencies to speed up compilation.
I also think that for an abstract class (an interface), `RooAbsMinimizer` has too many members and too many includes. Consider those suggestions optional, but if this code is touched, it could as well be modernised if time permits.

egpbos commented:
@hageboeck Thanks for the extensive review! I will implement some changes based on that and @guitargeek's comments today.

In the meantime, perhaps it would be useful to check out #8596 as well. That PR introduces the second concrete MinimizerFcn. Possibly, seeing `RooAbsMinimizerFcn` "in action" through that new class (`RooGradMinimizerFcn`) will shed a different light on the design choices for this abstract class. Basically, we previously just had two classes, `RooMinimizerFcn` and `RooGradMinimizerFcn` (and then later on **another** one, which will be introduced in another PR this week) with tons of code duplication. `RooAbsMinimizerFcn` simply factors that stuff out into one class.

guitargeek commented:
Hi @egpbos, FYI, there are conflicts in this PR now that https://github.com/root-project/root/pull/8609 got merged.

egpbos commented:
@guitargeek thanks for the heads-up, I will fix those later.

I have made a new commit, which addresses your and @hageboeck's review comments as much as possible. Point by point, let me quickly explain what changed:

- "Upgraded" comments on new functionality to doxygen. I did not doxygen document everything, because most functionality was copy-pasted from the old RooMinimizerFcn. The new things are now documented though.
- Renamed snake_case to dromedary.
- Use unique_ptr for parameter lists, as suggested by Stephan.
- Remove a few obsolete __ROOFIT_NOROOMINIMIZER flags (I guess these are part of the merge conflicts).
- I realized that part of what I changed in PR #8596 could (and imho should) have been introduced here already. I moved functionality to RooAbsMinimizerFcn, namely storage of the to-be-minimized function and constant term optimization information, but did not yet remove it (in this PR) from RooMinimizer. Now it has been removed.
- While going over the above, I realized I had actually *completely* forgot to add back in a backwards compatible way of enabling offsetting for numerical precision! This has now been correctly added back in.

Then, about what I did not change:

- I did not have time for doing the include reordering/optimization that @hageboeck suggested. I tried to quickly figure it out, but I couldn't change things fast enough and still be able to guarantee that everything will still work in the later PRs that build on this one. I agree it would be nice, but I hope we can postpone this to some future PR, perhaps after the other additions have been merged.
- As mentioned earlier, RooAbsMinimizerFcn is basically a way to reduce code duplication. This should be seen in light of RooGradMinimizerFcn as well (see again PR #8596) and will also be used for a later addition. I think removing all data members and also the protected vs private suggestion would complicate this goal.

I think it's good to merge like this (after fixing the conflict of course), but let me know if you have further suggestions!

egpbos commented:
Fixed the merge conflict as well now.

guitargeek commented on roofit/roofitcore/src/RooAbsMinimizerFcn.cxx:
```diff
@@ -0,0 +1,554 @@
+#include "RooMsgService.h"
+#include "RooMinimizer.h"
+#include "RooNaNPacker.h"
+
```
```suggestion
#include "TClass.h"
```

Because of the `arg->IsA()->InheritsFrom()` in the same file below. Unfortunately ROOT doesn't build on my system before this change.

egpbos replied:
Weird how those includes work differently on different systems sometimes... My IDE (CLion on macOS) even immediately notes that the include is unused. Perhaps something to do with platform-based compile flags then? Anyway, fixed!

guitargeek commented on roofit/roofitcore/inc/RooMinimizer.h:
```diff
@@ -109,9 +111,7 @@ class RooMinimizer : public TObject {
 
   Int_t       _printLevel ;
   Int_t       _status ;
-  Bool_t      _optConst ;
```
I only catch this now, but if you remove or add non-transient member variables to a ROOT class you have to update the class version. Which means that at the end of this file, you have update this:
```C++
ClassDef(RooMinimizer,0) // RooFit interface to ROOT::Fit::Fitter
```
to this
```
ClassDef(RooMinimizer,0) // RooFit interface to ROOT::Fit::Fitter
```

egpbos replied:
Ah good to know about the release cycle point, I indeed intended to bump the number in the final edits. Will do now instead.

guitargeek replied:
I see, good that you are keeping track of the class versions :+1: I just prefer to do at at the first edit in the release cycle so we won't forget.

guitargeek reviewed (changes requested):
Hi Patrick, I have two final requests: one to make the code build also on my system, and one related to the class version of `RooMinimizer` changing a ROOT class and forgetting to update the class version can have bad consequences for users, so it would really be better to do this already in this PR.

By the way, once the class version has been updated once in a given release development cycle, it doesn't need to be updated again in the same development cycle.

egpbos commented:
Ok, I made the two requested changes @guitargeek, thanks for the sharp reviews and support!

guitargeek reviewed (approved):
Looks good to me!
This PR splits up the RooMinimzerFunc into a RooAbsMinimizerFunc and an actual RooMinimizerFunc that inherits from it. This is to prepare for further RooFit developments. Most review comments have been addressed, for more details see the summary in https://github.com/root-project/root/pull/8569#issuecomment-875559598.

hageboeck commented:
@guitargeek @egpbos

Stop, stop!! Needs e3b73a8 to be reverted before releasing ROOT!

`ClassDef(..., 0)` means: Don't support I/O.
Now, you switched on I/O capabilities, so please revert! When people start writing minimiser states to files, you will have to support deserialisation in the future. It's better to not support that.

**That by the way also holds for any classes the are introduced in the MP-PRs. Please consider well wether they are allowed to support I/O or not.**

hageboeck commented:
@guitargeek If the class *had* supported I/O before, the situation would be even worse: In general
- You would need to write an instance of the class to a file, and save this file somewhere in the test suite.
- You would need to write a schema evolution rule to convert old data members to new ones (not in this case, since removing).
- You would need to add a test that reads back the old class instance, and tests that the new (in-memory) version is in a consistent state.

No harm done here, since you can just revert the one line, but please double check that none of the other PRs destroy I/O capabilities of RooFit.

egpbos commented:
Very sharp, thanks @hageboeck, I had no idea. Shall I do a revert PR for that one line?

egpbos commented:
Ah great, you already did, thanks!

egpbos commented:
Ah no, haha, it's an issue, ok, so shall I just make a PR to make the change or do you want to do it yourself? Is there some git procedure I should observe? I think `git revert` in this case is more work than just changing the one line.

hageboeck commented:
I just created an issue from the comment, so yes, please add a PR! 👍

guitargeek commented:
Ouch, that's painful! I totally fell for this, totally my fault. I suggested to increase the class number in the other PR. Now I will always use this as an example for why magic numbers are dangerous :D

Don't worry about this, @egpbos, I'll make a PR to revert the version number increase myself. I'll also add a comment to all RooFit classes with `ClassDef(..., 0)` so this mistake doesn't happen again.

egpbos commented:
Ah, damn, I didn't read your comment in time @guitargeek, I already made a PR as well. Please feel free to close. And I agree, magic numbers bad :D

egpbos commented:
Btw, if no I/O is necessary, is the ClassDef itself actually necessary?

Alternatively (if it is necessary): what about a macro overload without version number argument? And a static check in ClassDef that fails compilation on zero, referring to the new versionless one instead.

hageboeck commented:
It's necessary if you want ROOT's type inspection capabilities to work, e.g.
`this->IsA()->InheritsFrom(SomeThing::Class())`

In most cases, this doesn't have advantages over `dynamic_cast`, so I guess that you *mostly* want the ClassDef added when I/O is needed.

guitargeek commented:
Thanks Patrick for the PR!
